Checkout: add logged out checkout for Akismet

IanRamosC opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to 1201210512993648-as-1204136663449777

## Proposed Changes

* Add specific checkout flow for Akismet users who are not logged in. It allows users to complete the checkout process without a WordPress.com account

## Testing Instructions
* Follow the instructions in D105505-code
* Checkout this branch locally
* `yarn start`
* Go to http://calypso.localhost:3000/checkout/akismet/ak_plus_yearly_1 in a browser that doesn't have an account logged in (it can be the incognito mode)
* Add your email (e.g. your.email+test@gmail.com)
* Complete the purchase using a [test card](https://stripe.com/docs/testing#cards), and make sure you get the receipt and a welcome email

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

IanRamosC committed:
Checkout: add userless checkout for Akismet

IanRamosC commented on client/my-sites/checkout/composite-checkout/lib/submit-wpcom-transaction.ts:
```diff
@@ -25,11 +25,12 @@ export default async function submitWpcomTransaction(
-	const isJetpackUserLessCheckout =
-		payload.cart.products.some( ( product ) => product.extra.isJetpackCheckout ) &&
-		payload.cart.cart_key === 'no-user';
+	const isUserLessCheckout =
+		payload.cart.products.some( ( product ) => {
+			return product.extra.isJetpackCheckout || product.extra.isAkismetSitelessCheckout;
+		} ) && payload.cart.cart_key === 'no-user';
```
The change here is a little bit similar to the one in the file above, but I decided not to make a helper function because this is very specific and might vary from other places where we would want only one check or the other (i.e. if it has the product.extra properties, or the cart_key, or even just one of the properties).

sirbrillig commented on client/my-sites/checkout/composite-checkout/lib/create-wpcom-account-before-transaction.ts:
```diff
@@ -10,21 +10,31 @@ export async function createWpcomAccountBeforeTransaction(
 		( product ) => product.extra.isGiftPurchase
 	);
 
+	let signupFlowName;
```
Thanks! This is much better than the growing ternary it replaces. I think the following block is pretty readable right now, but in my experience, over time if/else code like this will get muddied and mixed up with other conditions. Could we use an IIFE instead? Something like this:

```suggestion
	const signupFlowName = ( () => {
	if ( isJetpackUserLessCheckout ) {
		return 'jetpack-userless-checkout';
	}
	if ( isAkismetUserLessCheckout ) {
		return 'akismet-userless-checkout';
	}
	if ( isGiftingCheckout ) {
		return 'gifting-userless-checkout';
	}
	return 'onboarding-registrationless';
} )();
```

This way the priority of each value is clear (the conditions do not overlap right now but I am sure they will eventually), assignments will clearly be made in one place (often when a variable can be modified folks will start making assignments in different areas over time) and the whole thing could be moved to its own function if needed.

IanRamosC replied:
Sure, that makes more sense to me as well. :)

sirbrillig reviewed (commented):
This code looks good to me!

However, when I tried to follow the testing instructions (and I'm probably doing something wrong) I get an error message:

<img width="786" alt="Screenshot 2023-03-22 at 11 07 39 AM" src="https://user-images.githubusercontent.com/2036909/226949472-7dda6cd8-2975-4ad6-b453-e0b7b1776d87.png">

Any advice?

IanRamosC commented:
> However, when I tried to follow the testing instructions (and I'm probably doing something wrong) I get an error message:

@sirbrillig, you're not doing anything wrong. I moved it back to In Progress yesterday because right before I committed to WPCOM, I rebased the branch I was working on - and trunk had code that made this PR stop working. I've updated the WPCOM diff, so this PR should work now.

IanRamosC committed:
Checkout: improve signupFlowName readability

The method createWpcomAccountBeforeTransaction contains an if that can eventually grow and get confusing. This commit refactors the code from if/else to an IIFE in order to make clear the priority of each condition, and improve its readability

jboland88 reviewed (approved):
Code changes look good and this tested well 👍 

I was redirected to a login screen after completing the checkout - it seems like this may be a result of the test setup. It looks like the expected behavior is to log the new user in 🤔 but we can test this more and follow up if needed.
